### ISSUE TITLE
Refactoring the Image file utilities 

### ISSUE THREAD

**Issue by kb-0311** [PalisadoesFoundation/talawa-api#952]
**Describe the bug**
Currently the Image file handling utilities for the application are not the best approach to solve the problem. 

As Specified in this comment ->
https://github.com/PalisadoesFoundation/talawa-api/pull/948#issuecomment-1407703633

I think instead of whatever we are doing now the best possible approach would be to

1. Convert the client side image to binary string and then make a mutation using that uri as string to the API.
2. In the API , we update the Object (Post ,User , Organization) image field with that binary uri .
3. Whenever we need to fetch or render images we can just do it by converting the Buffer binary data string of the object image field.
This way -

1. We really reduce the size of the data because we are not sending and retrieving MBs of png files back and forth between API and client.

2. Fixes this legacy flaw that [this whole fiasco revealed.](https://github.com/PalisadoesFoundation/talawa-api/pull/948#issuecomment-1407699700)

I will need to change the following files ->
1. `src/utilities/uploadImage.ts`
2. `src/utilities/reuploadDuplicateCheck.ts.ts`
3. `src/utilities/deleteImage.ts`
4. `src/utilities/imageExtensionCheck.ts`
5. `src/utilities/imageAlreadyInDbCheck.ts`
6. `src/utilities/deleteDuplicatedImage.ts`
In the API project and also make changes for talawa-admin and talawa mobile app repo so that changes can be made to the client side also.
**Screenshots**

**Additional details**

**Comment by kb-0311** [comment]
pls assign it to me

**Comment by kb-0311** [comment]
[The issue I am also currently working ](https://github.com/PalisadoesFoundation/talawa-api/issues/887)on can only be merged after these changes are merged . Thus PR will need to wait until this one is done

I will take around 3-4 days to complete the work for this particular refactoring issue.
Until then I will complete one more small testing (non-utilities) issue

**Comment by kb-0311** [comment]
@palisadoes pls checkout the comment under the now closed PR https://github.com/PalisadoesFoundation/talawa-api/pull/956#issuecomment-1410244587

**Comment by kb-0311** [comment]
@palisadoes - Update:
I have been learning about how to achieve plugins in the API and admin console and what I have come up with is ->
1. I will create a  module for the Integration of S3 in its own separate modular class in its own directory under `src/CloudProviders/Amazon-s3` where I would create subdir and files for the complete integration of S3 in the app with #764 as reference.
2. I will create a function to detect and load this plugin for the organization after creating the document under `Plugins` collection for the same to detect the existence of this plugin onload(). If and only if the plugin exists for the organization will the exported class from 1 be loaded as a functionality for the organization whose admin added the plugin and the image storage bucket will be s3 instead of the current filesystem.
3. if the plugin is not applied then the current way of using the filesystem will remain.
4. For the plugin to start , the organization admin  will need to [download it from the current Plugin Store](https://docs.talawa.io/docs/functionalities/plugin-architecture/#plugin-store) via the talawa-admin portal. After which he will need to pass the values for the env variables required for adding their s3 instance as a cloud provider to the backend and then trigger 2.

I would love to work on this issue but I have a humble request, as the passing of the environment variable will need encryption layer between talawa-admin and talawa-api for security reason ,

Which is a part of this idea - https://docs.talawa.io/docs/internships/internship-ideas/#api-improved-backend-performance-and-security 

I will be applying for GSoC this year under that idea and since the s3 integration will somewhat require the progression of that idea could you please add the integration proposal under that idea and make it a 350 hr project ? As then the GSoC period will have god time distribution after this proposal is added.

For now I will be writing the classes for integration mentioned in 1.

**Comment by kb-0311** [comment]
@palisadoes This feature could also be useful when we are giving org admins the option to store their binary data on their specific cloud providers. In other words, this approach could easily open up to develop the options to provide multiple cloud storage providers for that.

What do you think about the proposal in the above comment ?

**Comment by xoldyckk** [comment]
@palisadoes @kb-0311 You would be introducing needless complexity by integrating s3 into the project if all you need is a way to store small images. Mongodb offers an inbuilt storage solution known as [gridFS](https://www.mongodb.com/docs/manual/core/gridfs/). It would be easier to use just this.

Though if you want better latency, high throughput etc., or the files you're storing are too large like hundreds of megabytes or even bigger then s3 is the viable choice. Besides s3 comes with additional complexities which would a pain to set up for new contributers and it's not a free service.

**Comment by kb-0311** [comment]
@xoldyckk Thanks for your feedback  - 
1. The integration of s3 was not meant as a complete replacement of the file uploading system in the proposal but rather a plugin for admin-portal based on the choice of the organization admin as a choice for a cloud provider. 
2. Great suggestion. I was just thinking about ways to store binary files in the database and stumbled upon gridFS. It could serve as a primary approach to completely replace the current  file system storage, which is great since there is a limit to how many files there can be in a directory. 

So just to summarize - 
1. integrating gridFS can serve as a complete replacement for file management storage for the API.
2. (Optional) Integrating s3 as a plugin and not a part of the main API functionalities to go through with it.

I think your suggestion is great , you could be onto something here,
@palisadoes @xoldyckk should I start working on replacing the file system storage    with gridFS this week ?

**Comment by xoldyckk** [comment]
@kb-0311 lol I didn't read your proposal sorry. The whole idea of using plugins for integrating different storage providers seems really solid and future proof. For storing static files on local file system a native inbuilt solution like gridfs would be good.

Btw the whole round trip of retrieving the files from s3 then sending them to clients is wasteful. It would be better for the clients to directly request the resources from s3. The api just needs to send back the url to retrieve the resource from s3 bucket or gridfs. Though idk what the complications of that approach are. Just wanted to add my opinion.

**Comment by kb-0311** [comment]
@xoldyckk 
>The whole idea of using plugins for integrating different storage providers seems really solid and future proof. For storing static files on local file system a native inbuilt solution like gridfs would be good.

Thanks a lot for your compliment, means a lot. 

>Btw the whole round trip of retrieving the files from s3 then sending them to clients is wasteful. It would be better for the clients to directly request the resources. The api just needs to send back the URL to retrieve the resource from s3 bucket or gridfs. Though idk what the complications of that approach are. Just wanted to add my opinion.

Yes and that's why retrieval of files from the s3 bucket was never my idea, my idea was - when an admins virtually deployed s3 instance is configured in the installed plugin, then a file uploaded will first be stored in s3 bucket and deployed on as a  website serving files . The image then would be in the form of a virtually hosted URL when an image - `puppy.png` is uploaded ->

`https://my-website.s3.us-west-2.amazonaws.com/puppy.png`

So basically we store this URL as string for image field for USER/ORGANIZATION etc. In  the client side while rendering ,  this URL from the CDN is fetched and image is rendered directly instead of file retrievals.

(ofc we will need to hash the image names first)

**Comment by palisadoes** [comment]
- The uploaded content should only be made available to people of the same organization. How would you ensure this would happen with cloud providers? It's an area that I'm not that familiar with.
- Also we have to remember that most non-profit organizations may not have the funds to pay for cloud hosting, so storage on the API server will need to be the default plugin option.

**Comment by kb-0311** [comment]
> * The uploaded content should only be made available to people of the same organization. How would you ensure this would happen with cloud providers? It's an area that I'm not that familiar with.

I have a solution for this but it is a lot complicated and not worth exploring rn , let's just hold on to the whole s3 idea for a minute and allow me to discuss about the other point - file storage on file system on the API server. 


> * Also we have to remember that most non-profit organizations may not have the funds to pay for cloud hosting, so storage on the API server will need to be the default plugin option.

It seems that we are using Heroku as our backend deployment. Why storing images in file system is  terrible idea - 
1. The deployment has ephemeral disk that means all the files written on its file system gets deleted once every 24 hours during cycling or every time a new deployment is published. Source -https://devcenter.heroku.com/articles/active-storage-on-heroku#ephemeral-disk
2. Even if we get a work around for that there is the issue of the size of code that can be deployed on production , for heroku the slug size is only 500MB , for image files across multiple orgs , groupchats , direct chats , users etc this is really low . After the slug size becomes > 500MB it throw R15 error how are we planning to scale in this situation? Source -https://devcenter.heroku.com/articles/slug-compiler#slug-size 

4. Even if we ditch heroku, and move on to some other type of deployment. Lets say the above two problems are solved but in the event when a deployment is failed and we need to redeploy the code , then all the images will be lost. How are we planning to back this up in this case ?
  
I strongly propose implementing @xoldyckk suggestion for now about using MongoDB's inbuilt gridFS and storing the image in a separate database, this will make backups , horizontal scaling and separation of concern much easier.

As @xoldyckk said - 
>Mongodb offers an inbuilt storage solution known as [gridFS](https://www.mongodb.com/docs/manual/core/gridfs/). It would be easier to use just this. S3 comes with additional complexities which would a pain to set up for new contributers and it's not a free service.
Though if you want better latency, high throughput, less storage charges etc., or the files you're storing are too large like hundreds of megabytes or even bigger then s3 is the viable choice.

**Comment by palisadoes** [comment]
Please provide some more detail on the MongoDB based proposal. For example:
1. What new settings will be required?
2. What minimum system requirements will be needed?
3. Etc.

**Comment by xoldyckk** [comment]
@kb-0311 This is assuming talawa-api is deployed on heroku right?

**Comment by palisadoes** [comment]
We need a solution that will work first on a local installation of Mongo, then on a hosted version. Our documentation will need to be updated to reflect the different scenarios.

**Comment by xoldyckk** [comment]
@palisadoes The whole point of using gridFS is to have a remote backup of static files because heroku has ephimeral file system. It doesn't provide a permanent storage solution like s3. So, once the dyno is restarted all the static files uploaded by talawa users is wiped. There's also the issue of how many files you can store on a native file system.

GridFS does work with local installation of mongo. Correct me if I'm wrong @kb-0311. So, no problem for contributers.

**Comment by xoldyckk** [comment]
> 

Didn't understand point 4. You're going to cache static files on the file system and when it's full you'll do a clean wipe and start caching from 0? Really weird way of doing it I'd say.

**Comment by xoldyckk** [comment]
> > Please provide some more detail on the MongoDB based proposal. For example:
> > 
> > 1. What new settings will be required?
> > 2. What minimum system requirements will be needed?
> > 3. Etc.
> 
> Sure, Sir.
> 
>     1. The gridFs images will be stored in the same database as other metadata. Grid fs will creat two collections - fs.chunks(for storing file data) and fs.files(for storing image meta data like file name , date etc). We can access the stored files based on file name which will be hashed when an image is uploaded. All the logic for reupload duplicate check will remain relatively unchanged.
> 
>     2. I went ahead and created a sample code snippet for mongoose (which is a work in progress)- https://github.com/kb-0311/talawa-possible-gridfsSoln
>        Basically, after a file has been uploaded successfully, we can retrieve it using its file name from the bucket , then we will write its data temporarily on the file system. Think of it as a backend cache for image files. Sooo before retrieval, we will check whether image file with the same filename exists or not in the ephemeral file system if yes then we send back the file if not then we will make a round trip to the database and fetch the image and write it into the file system.
> 
>     3. This way we will get the optimization of file system along with conistency and scalability of database,  and even if the ephemeral disk is cleared we have all the image data stored in db .
> 
>     4. before a file is uploaded we will check if the images/ dir which will serve as out file system cache is within the size of 300MB if it surpasses that then the image dir will be deleted.
> 
>     5. As gridFS is a mongoDB solution, the system requirements will be the same as mongodb as of the change in the project setting, right now I don't think we will require any as mongoose itself has its own GridFSBucket under its mongo object.
> 
> 
> So, what do you think ?
> 
> @xoldyckk btw do you have any suggestions regarding this one something left uncovered above ?

Didn't understand point 4. You're going to cache static files on the file system and when it's full you'll do a clean wipe and start caching from 0? Really weird way of doing it I'd say.

**Comment by palisadoes** [comment]
I agree that we should not use the local file system, however I meant a non-cloud Mongo installation, such as an on premises solution.
We have killed the Palisadoes Heroku instance because of maintainability concerns. I didn't realize that Heroku files were ephemeral. This is good to know and a benefit of making that decision and proposing this one. 

So it sounds like Heroku is best for temporary caching of content objects.

**Comment by kb-0311** [comment]
@xoldyckk yes , gridFS also works with local instances of mongo so now extra hassle for contributors.

regarding my point 4, it is necessary to because we have to make sure  that the dynamic size of the API server slug does not exceed 500MB because of point 2 here -https://github.com/PalisadoesFoundation/talawa-api/issues/952#issuecomment-1413395447

**Comment by xoldyckk** [comment]
> 1. I agree that we should not use the local file system, however I meant a non-cloud Mongo installation, such as an on premises solution.
> 
>     2. We have killed the Palisadoes Heroku instance because of maintainability concerns. I didn't realize that Heroku files were ephemeral. This is good to know and a benefit of making that decision and proposing this one.
> 
> 
> So it sounds like Heroku is best for temporary caching of content objects.

@palisadoes So, you want the static files to stay close to the api server?

**Comment by kb-0311** [comment]
@palisadoes @xoldyckk 
I had already written all the points why maybe storing image files in local API server is not a good idea here https://github.com/PalisadoesFoundation/talawa-api/issues/952#issuecomment-1413395447
here I had also mentioned ephemeral disk for Heroku but the thing is most of the cloud deployers have ephemeral disk along with a slug size limit. having a non - cloud solution for this would be tough to implement presuming there is one.

**Comment by palisadoes** [comment]
>     4. There is ofc the event of some idiot pushing code into the main branch that could delete these files' directory . Source - [urgent bugFix #948](https://github.com/PalisadoesFoundation/talawa-api/pull/948) (lol). The files will be deleted without any backup.

The purpose of testing is to catch these types of errors. We all make mistakes. The identification of the error was swift as was the recovery. Making the error a second time would be truly careless. Let's maintain our community values of a welcoming environment.

**Comment by kb-0311** [comment]
@palisadoes yeah that point was only meant as a joke lol

**Comment by palisadoes** [comment]
> @palisadoes yeah that point was only meant as a joke lol

I just realized that it was self deprecating. Sorry.

**Comment by xoldyckk** [comment]
@palisadoes she was calling herself an idiot. I think you took it the wrong way.

**Comment by xoldyckk** [comment]
> > 1. I agree that we should not use the local file system, however I meant a non-cloud Mongo installation, such as an on premises solution.
> >    
> >    2. We have killed the Palisadoes Heroku instance because of maintainability concerns. I didn't realize that Heroku files were ephemeral. This is good to know and a benefit of making that decision and proposing this one.
> > 
> > So it sounds like Heroku is best for temporary caching of content objects.
> 
> @palisadoes So, you want the static files to stay close to the api server?

@palisadoes this is what you meant right?

**Comment by kb-0311** [comment]
@xoldyckk I hope I cleared up your doubt on point 4 in the previous comment

**Comment by palisadoes** [comment]
> > @palisadoes So, you want the static files to stay close to the api server?
> 
> @palisadoes this is what you meant right?

1. Our INSTALLATION.md file focuses on setting things up on a local server (Mongo or otherwise), so yes, the default installation would have the files close to the API server.
1. We must also show the benefits of the remote serving too in INSTALLATION.md

We have to consider two types of user. Developers setting up on their local system and administrators who may or may not want to do that.

**Comment by xoldyckk** [comment]
@palisadoes It's not about the benefits. It's about having a permanent storage(like s3) which has a backup of all the static media files uploaded by talawa users/orgs. Heroku doesn't provide a solution like that. So, if you want a on premise solution you have to move to a different cloud provider which does provide it. If you're hosting on heroku you have to use a remote storage solution from other other provider.

**Comment by kb-0311** [comment]
Yes, let me just elaborate on both of these points that you raised - 
1. gridFS can be worked on with local instances on MongoDB as well as remote instances. All it does is add two extra collection  in developers' db - `fs.chunks` and `fs.files` . So the initial setup / configuration is practically unchanged. 
2.  And in development I have already specified how it will work here https://github.com/PalisadoesFoundation/talawa-api/issues/952#issuecomment-1416253564

**Comment by palisadoes** [comment]
> @palisadoes It's not about the benefits. It's about having a permanent storage(like s3) which has a backup of all the static media files uploaded by talawa users/orgs. Heroku doesn't provide a solution like that. So, if you want a on premise solution you have to move to a different cloud provider which does provide it. If you're hosting on heroku you have to use a remote storage solution from some other provider.
> 
> Unless the org admin does not consider having a backup of their static files a benefit and only want to store their static files temporarily which is wiped each day?
> 
> And of course there's the issue of 500mb limit on heroku.

We can discuss this on our call tomorrow. I lack the background, but really want to understand the long term implications. I think there are unspoken assumptions that a verbal discussion will rectify.

**Comment by kb-0311** [comment]
@palisadoes will it be okay If I could participate in a verbal tomorrow too? Because I have explored a lot of alternative solutions to tackle this problem and how to go about implementing gridFS if it could be implemented that is.

**Comment by xoldyckk** [comment]
@kb-0311 It'll be you of course it's your idea. Besides I'm only providing input I don't exactly know the nitty gritty details of of the implementation.

**Comment by kb-0311** [comment]
> @kb-0311 It'll be you of course it's your idea. Besides I'm only providing input I don't exactly know the nitty gritty details of of the implementation.

Thanks a lot , Sir.

Umm btw I am relatively new here how do verbal calls in the organization work exactly ? Like a google meet or something?

**Comment by palisadoes** [comment]
@kb-0311 Please create a channel on slack so that we can coordinate.

**Comment by palisadoes** [comment]
@kb-0311 , @xoldyckk 
1. We need to get basic file upload functionality for the API so that the Mobile app and Admin can use the feature.
2. Hopefully we can do this without requiring new GraphQL endpoints. If we have to change it, then we'll need to let everyone know with some anticipation

What would you consider basic functionality that could be implemented within a week? I'd greatly prefer to have this done before the GSoC organization announcements.

**Comment by kb-0311** [comment]
@palisadoes Yes I will get the basic file uploading, storage, and deleting in the database done for the contributors next. But as we discussed earlier on call I am a little preoccupied with my university exams , which will end on 15th Feb so till that time I will be working on this locally little by little since it will require complete attention on my part to not cause breaking changes.

However, I do promise that as soon as my exams are over on 15th Feb, I will dedicate most of the time to GridFS implementation. GSoC org announcements are on 22nd Feb, I will try to complete at least the basic Image object handling by that time. Till then 15th, I will submit PRs for more minor issues here and there.

**Comment by kb-0311** [comment]
Progress (1/4) : 
In the call you mentioned, there needs to be a method to make sure the same exact binary object does not get uploaded again and again taking up space. So for that , it turns out there was a sort of solution here already in this file, -> 
https://github.com/PalisadoesFoundation/talawa-api/blob/develop/src/utilities/reuploadDuplicateCheck.ts

Basically, if I am not wrong we are hashing the binary of an image and storing the hash value in `ImageHash` model and then instead of reuploading the same file again and again we make sure the image field points to the already uploaded binary object in the database. 

This one is a good utility in my opinion What do you think?

**Comment by kb-0311** [comment]
@palisadoes @xoldyckk Good news:My exams are over and I am working on this issue full time now.

Bad news:

The current apollo server implementation in the project is wholly deprecated (by that I mean even if we upgrade it to v3, v3-is-going-to-be-deprecated-on-October-this-year-kind deprecated-deprecated). Leaving the graphql-upload functionality (the basis of file uploads in our server) open to CSRF attacks.(In the current Project that we are using apollo server v2 which is open to CSRF attacks even WITHOUT file uploads integrated as it has a very deprecated older version of `graphql-uploads` package integrated by default , leaving our production server CSRF attacks at anytime)

https://www.apollographql.com/blog/graphql/file-uploads/with-apollo-server-2-0/

 In order to completely prevent CSRF attacks and integrate make the file uploads functional we will first need to ->
1. upgrade apollo server to v3 then to v4 asap (v3 is also deprecated!). Apollo v4 has a good support for integrating Upload scalar and inbuilt function to prevent these CSRF attacks
2. integrate `graphql-upload` package individually.
3. pass in the `csrfPrevention:true` config to the apollo server constructor and set upload scalar with graphql-upload file.
4. Configure the `Apollo-Require-Preflight` headers in the client side too.

Which begs some questions.
1. Was the current upload image functionality ever properly tests? I don't mean unit tests like actual proper end to end tests with the client side (admin portal or mobile app ? Since even the current file upload version does not work with client-side integration at least while I am working on it. 
2. We will need to migrate the project to apollo server v4 before October 2023. Which should not be completely hard since according to apollo docs it requires almost no code changes from v2->v3->v4 asap if we need `graphql-upload` to work? Is there any reason for this delay?
3. If yes, how soon do you want the Image object uploads to be functional? Because as I said without the migration, it just does not work.
4. I am willing to raise a separate issue for the migrations and work on this and make sure that at least the basics of file uploads is done before 22nd Feb as we agreed BUT it will not be fair to @EshaanAgg who had raised the issue of migration before . I am just giving more reasons to do so here.

https://www.apollographql.com/docs/apollo-server/previous-versions/#apollo-server-2

**Comment by xoldyckk** [comment]
@kb-0311 If you ask me I'd say we should completely ditch apollo-server and instead make use of [graphql-yoga](https://the-guild.dev/graphql/yoga-server/docs/comparison). It's built on web standards, supports all modern features, less bloated compared to apollo-server, keeps itself minimal and let's the user choose what they want to use.

It's also officially maintained by [the-guild](https://the-guild.dev/#platform). They're the ones responsible for maintaining most of the well known packages in javascript graphql ecosystem. 

With graphql-yoga you can make use of a plugin architecture to make use of the plethora of [envelop graphql plugins](https://the-guild.dev/graphql/envelop/plugins) to integrate different functionalities in an easy way.

Migration from apollo-server to graphql-yoga is also very easy. Look at [this](https://the-guild.dev/graphql/yoga-server/docs/migration/migration-from-apollo-server). Not much to do at all.

I'd also like to say we should replace many other packages used by talawa-api to their modern counterparts which have active support, work better with typescript, are more performant etc.

**Comment by kb-0311** [comment]
@xoldyckk I agree with you however, there is a soon becoming a long list of changes to make to the project dependencies as discussed in other threads (Prisma ORM , enforce graphql-pattern into queries etc )which probably needs to be done one by one.

Sure the migration part is easy but it does not support batching. Also to perform functions available in the apollo server it requires external packages as plugins. (for eg for CSRF attacks it requires another plugin etc). Not saying it is not worth it but I doubt I will be able to complete it before the 22nd.

Secondly, migrating to apollo v4 for at least the short term to make sure the images upload/delete, etc are functional on the client side does make me think that at least for now we should do minimal changes to the dependencies. SO migrating to v4 does seem like a more feasible option at least for now in short term.

**Comment by EshaanAgg** [comment]
@xoldyckk @kb-0311 
I had started working on the migration from `apollo-server-2` to `apollo-server-3`, but found there were many breaking changes in the same, which required refactoring the same. Seeing the same, @palisadoes decided to close the issue then, and keep it as a `GSoC` proposal idea.

In view of using `graphql-yoga`, I am personally in complete support of the same. But before making this switch, I feel that we should leave this discussion open for a couple of days so that other active developers can also drop their views about the same. We will need also additional review on the fact, that should we migrate the project on a `couple of PRs that migrate the whole codebase` basis, or as a complete project in itself basis. 
Testing of image utils should only be worked upon once we arrive at a stable server for `talawa-api`.

**Comment by kb-0311** [comment]
@EshaanAgg @palisadoes duly noted. So for this PR I will be at least making sure the underlying API logic for gridFS integration in done .

The client side integration will ofc be after the migration.

@EshaanAgg I take it you will be working on migration project for GSoC? How long do you think it will take for you to complete it ?

**Comment by kb-0311** [comment]
@palisadoes could you please add the apollo server migration idea to the talawa-docs please? thanks.

**Comment by EshaanAgg** [comment]
> @EshaanAgg I take it you will be working on migration project for GSoC? How long do you think it will take for you to complete it ?

I would definitely try to take it up during that time! As to comment on the timeline for the same, I'll need to read the documentation of both the current and the proposed server in detail, which is something I'll get to most probably in a week's time (as I have my semester exams going on right now).

**Comment by xoldyckk** [comment]
> @xoldyckk @kb-0311 I had started working on the migration from `apollo-server-2` to `apollo-server-3`, but found there were many breaking changes, which required refactoring the same. In view of this, @palisadoes decided to close the issue, and keep it as a `GSoC` proposal idea.
> 
> In view of using `graphql-yoga`, I am personally in complete support of the same. But before making this switch, I feel that we should leave this discussion open for a couple of days so that other active developers can also drop their views about the same. We will need also additional review on the fact, that should we migrate the project on a `couple of PRs that migrate the whole codebase` basis, or as a complete project in itself basis. Testing of image utils should only be worked upon once we arrive at a stable server for `talawa-api`.

It does [support batching](https://the-guild.dev/graphql/yoga-server/docs/features/request-batching), just not enabled by default as batching is a bad practice.

CSRF is a [one liner](https://the-guild.dev/graphql/yoga-server/docs/features/csrf-prevention) to integrate.

Schema directives(like @auth) in apollo-server 4 have a new implementation. They're not the same as talawa-api currently uses. They make use of graphql-tools which again is another package maintained by the guild. So, you will have to code it from scratch if you're gonna use apollo-server 4.

Btw I think you didn't go through their docs properly so you don't see the value it would provide. There are many other useful features it provides.

Whatever effort it'll take to migrate to apollo-server the same will apply to graphql-yoga too. So, if you're going to put effort in something you should choose wisely and think of the future prospects.

Most of the graphql related code lives in resolvers and typeDefs which don't depend on what server implementation you use for graphql. So, they won't need to be changed.

**Comment by xoldyckk** [comment]
I personally think it's a bad idea to include upgrading packages and dependencies as a GSOC idea. This is something which should be continually maintained in the project instead of remaining stale and waiting for a certain event(like GSOC) to be fixed.

This will only create friction between features implemented by different contributers as there will be version dependency conflicts. If you have the time for it right now why wait for later?

**Comment by kb-0311** [comment]
@xoldyckk I too would prefer that migrations would be done before the GSoC period starts so that this image thingy could be implemented before that, no point in delaying it. 

However, I think it will take a little more time than what was previously agreed - (22nd feb)

**Comment by palisadoes** [comment]
The aim of upgrading packages as part of GSoC was to make sure that they were all up to date by the end of the period. They would then be kept current with dependabot.

It wasn't the intention to stop upgrades completely until then.

If there was no clear path to doing this upgrade, then it should be done as an idea.

I'll upgrade the wording of the idea

**Comment by palisadoes** [comment]
I support the upgrade or replacement of this package if it isn't too disruptive.

**Comment by kb-0311** [comment]
@palisadoes so basically the migration as GSoC period was only meant as a plan if there was no activity from contributors before the period to do so or if the migration was incomplete, right?

So basically if @EshaanAgg and I coordinate along with other contributors on wanting to migrate the apollo server to `graphql-yoga` by working on it on a separate branch before the GSOC point it is okay, right?

**Comment by EshaanAgg** [comment]
I'll be willing to contribute to the same passively in the upcoming days, and then actively from 22nd Feb, after my exams end.

**Comment by anshgoyalevil** [comment]
I will also be helping in migrating Appollo Server to GraphQL -Yoga.

I will be having my exams till 27 Feb, so I would be less active until then. After that, my whole time would go to contributions.

**Comment by kb-0311** [comment]
@EshaanAgg @anshgoyalevil ofc that would be great. 

I don't have any exams for now and can contribute actively. I will be creating a branch in my forked repo for migration where I will be beginning with at least the starter files. you two can check out that PR and then after your exams are over you can push commits to that branch. 


Should I go ahead and create a slack channel for the same?

**Comment by palisadoes** [comment]
@kb-0311 @EshaanAgg @anshgoyalevil @xoldyckk Are there any other packages maintained by [the-guild](https://the-guild.dev/#platform) that we should consider as replacements for what we already have?

**Comment by xoldyckk** [comment]
> @kb-0311 @EshaanAgg @anshgoyalevil @xoldyckk Are there any other packages maintained by [the-guild](https://the-guild.dev/#platform) that we should consider as replacements for what we already have?

Talawa-api already is using many of their tools like @graphql-tools, @graphql-codegen, graphql-eslint. For replacement there's only apollo-server to be replaced. For enhancing what's already present there's scalars, envelop and hive/inspector.

All other tools are focused on developer productivity and microservices architecture.

**Comment by xoldyckk** [comment]
> @EshaanAgg @anshgoyalevil ofc that would be great.
> 
> I don't have any exams for now and can contribute actively. I will be creating a branch in my forked repo for migration where I will be beginning with at least the starter files. you two can check out that PR and then after your exams are over you can push commits to that branch.
> 
> Should I go ahead and create a slack channel for the same? OH, and @xoldyckk would it be cool if I add you there, too?

Yeah sure

**Comment by SiddheshKukade** [comment]
> @palisadoes - Update: I have been learning about how to achieve plugins in the API and admin console and what I have come up with is ->
> 
> 1. I will create a  module for the Integration of S3 in its own separate modular class in its own directory under `src/CloudProviders/Amazon-s3` where I would create subdir and files for the complete integration of S3 in the app with [Backend support api ready along with AWS S3 connection. #764](https://github.com/PalisadoesFoundation/talawa-api/pull/764) as reference. This directory structure will allow us to develop feature to integrate multiple cloud storage providers as plugins in the future.
> 2. I will create a function to detect and load this plugin for the organization after creating the document under `Plugins` collection for the same to detect the existence of this plugin onload(). If and only if the plugin exists for the organization will the exported class from 1 be loaded as a functionality for the organization whose admin added the plugin and the image storage bucket will be s3 instead of the current filesystem.
> 3. if the plugin is not applied then the current way of using the filesystem will remain.
> 4. For the plugin to start , the organization admin  will need to [download it from the current Plugin Store](https://docs.talawa.io/docs/functionalities/plugin-architecture/#plugin-store) via the talawa-admin portal. After which he will need to pass the values for the env variables required for adding their s3 instance as a cloud provider to the backend and then trigger 2.
> 

> 

> For now I will be writing the classes for integration mentioned in 1.

@kb-0311  I know I'm pretty late to take a look at this but the point no 02 is already implemented by me.

**Comment by kb-0311** [comment]
@SiddheshKukade Yep. No worries though the entire point about plugins is out right now . https://github.com/PalisadoesFoundation/talawa-api/issues/952#issuecomment-1414326730 this is more of a 'recent' thingy

**Comment by palisadoes** [comment]
1. This will need to be fixed with a long term solution. 
1. For our immediate purposes, please consider a way to generate random JPGs as part of the API response. We'll use GSoC for a long term fix.
1. The lack of this feature is blocking development in many other areas.

**Comment by palisadoes** [comment]
The short term solution is being worked on here: https://github.com/PalisadoesFoundation/talawa-api/issues/1080

**Comment by palisadoes** [comment]
Earlier today, during the Talawa functionality review meeting with GSoC candidates GridFS came up again for storing images. During the discussion it was stated that:

1. GridFS works best with files larger than 12MB.
1. Some form of encoding could be used to store images in the DB with performance degradation.

What if we combined the two? How about:

1. Concatenating all images in a single file stored in GridFS
1. Creating a second or DB entry pointing to the start points for each image with some additional metadata like MP4, JPG, PNG etc.

@DMills27 @Kevoniat @noman2002 what do you think?

**Comment by kb-0311** [comment]
@palisadoes 
Only for the purpose of development to make this feature functional as we are on a time crunch right now, I am okay with full encoding/decoding images and storing it as a string so that the new contributors for mobile app and admin portal can use this functionality asap. I will implement that in #1080 and submit a PR for it at night(in my timezone)

But just so we are on the same page, I would strongly suggest against any kind of decoding or encoding of binary data in production or in the later stages of GSoC. 
I had explained my reasoning here while I was trying to do exactly that. ->

https://github.com/PalisadoesFoundation/talawa-api/pull/956#issuecomment-1410244587

Furthermore, I had made it a point to ask whether video file(which are generally >16MBs) will be a feature in the mobile app , where the answer was yes. If we will be following this approach we will also have to encode/decode video files(oof). 

Also I had worked with this approach earlier in one of my hobby projects in react (the UX fin production or image upload was _really_ poor)

https://github.com/kb-0311/ConsMatrix/blob/654b77a1caa7806b0fe5963a3a75ee659a340ebf/client/src/components/User/UpdateProfile.js#L37

**Comment by palisadoes** [comment]
I'm not saying that we do it. It just crossed my mind.

It could be one way to solve the problem in the long term and would require investigation. 

There may be others of course.

**Comment by palisadoes** [comment]
I'm not saying that we do it. It just crossed my mind as a way of using a single methodology for all types of object storage.

Whether string or binary data, the approach may be usable.

It could be one way to solve the problem in the long term and would require investigation. 

If the metadata included a hash of the image data, then deduplication could be easier.

There may be other ways of course.

**Comment by kb-0311** [comment]
@palisadoes Btw should I close it for now? Since we have already decided it is best to have it as a GSoC Idea this year?

**Comment by palisadoes** [comment]
1. Now that we are at the latest version of Apollo, should we start considering migrating to `graphql-yoga`?
1. I remember there was a decision to defer the work until we got the API more stable and Apollo upgraded. @EshaanAgg (or maybe it was someone else) tried and determined it was too much work starting with Apollo V3 which we had at the time of this PR

**Comment by disha1202** [comment]
> 1. Now that we are at the latest version of Apollo, should we start considering migrating to `graphql-yoga`?
> 1. I remember there was a decision to defer the work until we got the API more stable and Apollo upgraded. @EshaanAgg (or maybe it was someone else) tried and determined it was too much work starting with Apollo V3 which we had at the time of this PR
> 
> 

Hi @palisadoes 
If we're planning to migrate to graphql yoga. I would really like to take this up.

**Comment by varshith257** [comment]
> 1. Now that we are at the latest version of Apollo, should we start considering migrating to `graphql-yoga`?
> 2. I remember there was a decision to defer the work until we got the API more stable and Apollo upgraded. @EshaanAgg (or maybe it was someone else) tried and determined it was too much work starting with Apollo V3 which we had at the time of this PR

@palisadoes 

1. Yes we can start with it, I have followed all discussions regarding discussions about migrating of Apollo Server  and @xoldyckk 's suggestions to migrate for graphql-yoga instead latest versions of Apollo. 
2. IMO, the API is now stable and we can surely start migration to graphql-yoga before GSOC contributors start work on their projects after selection. If it is delayed now maybe it can lead to the unexpected complexities of migrating to graphql-yoga.

@xoldyckk Your suggestions on it. I have followed all the links suggested by you for migration and got familiar with the  Apollo server migration to graphql-yoga. Could create an issue and start working on migrations?

**Comment by varshith257** [comment]
@xoldyckk Waiting for your suggestions here  :)

**Comment by xoldyckk** [comment]
Replacement with graphql-yoga is very easy, there's not much to it. Essentially the only part being changed is the graphql engine.
 
Expressjs and related packages are also mostly useless and a liability since `/graphql` is the only route being exposed from talawa api. So, most of the fluff related to it can also be removed.

If there really is a need for backend framework, some well maintained and modern framework like fastify or hono should be used. An example use case would be authentication related mechanisms like oauth openid connect which has to be handled using REST api based endpoints.

**Comment by xoldyckk** [comment]
Current code logic for authentication mechanisms(sign in, sign up, sign out) is naive and full of security risks. It would be better to offload this work to to a library meant for this([lucia-auth](https://lucia-auth.com/), passport.js etc) or a third party provider all together(okta, auth0, clerk etc). Rolling out a production level authentication system from scratch is incredibly hard and not something that should be wasted time on.

Authorization logic is scattered over the codebase and is done imperatively from scratch whereever it is needed. It needs to be centralized in one place and be made use of wherever needed in the codebase. The authorization decisions also have to be well documented, it is very tedious and error prone to just remember it. [Oso](https://www.osohq.com/docs/oss), [casl](https://casl.js.org/) [casbin](https://casbin.org/) etc., are a libraries that come to mind for this task. Casl in particular is easier to get started with because it is just javascript, but fails when the authorization logic is complex.

**Comment by varshith257** [comment]
Great suggestions @xoldyckk. Moving our backend framework from express to modern frameworks like fastify or hano adheres to current industry practices and adapting new technologies and tools to our project. Let's go step by step implementing the features you suggested the above suggestions are really amazing for incorporating the latest technologies into our project. 

Let's go with Fastify for the backend framework as its main aim is to support Typescript community and which lacks the same in hano. Fastify has LTS and has shown to be significantly faster than Express in terms of request throughput and response time and has same features as Express support

I am about to talk about this current auth mechanism in the project. I thought it could be done after GSOC but the suggestions made by you seem like a viable solution incorporating authentication and authorization into the project. I think we can use the **passport.js** which is widely using the library to offload this authentication work instead of relying on third-party providers. 

For authorization let's go with Casl as it provides the flexibility of implementing in frontend and backend, we can use  `casl` in Talawa-Api and use the `packRules `and `unpackRules `function in` @casl/ability/extra` to transfer the necessary rules to Talawa-Admin. We can go with standard authorization logic that also supports security features and is currently the most used in the industry.OSO integration makes a larger learning curve and learning polar and difficult in integration.

**Comment by palisadoes** [comment]
1. What would the list of issues to create be and for what repos?
2. I'd prefer to apply them to the `develop-fixUserType` branch as that fixes the user permissions for Admin, SuperAdmin and regular Users

**Comment by varshith257** [comment]
Sure! We can apply them to the `develop-userTypeFix` branch. Although we decided to use **Fastify** instead of express, let's start first with it because if we start graphql-yoga migration, it has to be changed again to support fastify. 

Instead of making all issues at a time and working parallel, I'd prefer to integrate the above features one by one to ensure compatibility with each other. They have to wait for each other to get merged as a dependency of other issues if we start working parallelly. Once fastify is restored we can keep track of these features and integrate them easily without any dependent work for them. I just started yesterday looking at the migration of express to Fastify and will create an issue and work on it.

**Comment by varshith257** [comment]
@palisadoes Great to have a plan in place!  Here's a breakdown of the major steps and sequencing for our above tasks:

**Fastify Integration:**

- Major Steps: 

1. Setup Fastify as our backend framework instead of express server as it is a modern web framework and has  LTS for TypeScript Community and a widely used framework and also has integrated logging feature
2. It will require major changes in setup files like `index.ts`, `app.ts` etc... Will require Incremental migration from Express to Fastify to restore functionality as same as the express and to ensure compatibility.

 References:
https://fastify.dev/docs/latest/Reference/TypeScript/
https://blog.appsignal.com/2023/06/28/migrate-your-express-application-to-fastify.html

**GraphQL-Yoga Integration:**
- Major Steps: 

1. Once fastify is restored as the backend framework, we can easily integrate graphql-yoga as @xoldyckk mentioned with related changes of graphql-yoga with the current Apollo Server and with fastify integration of graphql-yoga. 
2. Remove all functionalites and packages of express. If needed we can use modern alternative packages

References: 
https://the-guild.dev/graphql/yoga-server/docs/integrations/integration-with-fastify

**Comment by varshith257** [comment]
*While authentication and authorization are logically separate concerns, they often depend on each other. Authentication is a prerequisite for effective authorization. Coordination between the authentication and authorization aspects is crucial.**

**Authentication**

Passport.js is a great choice for our project because it is easy to use and provides a lot of flexibility. It is also well-maintained and has a large community of users. Passport.js supports a variety of authentication strategies, including local username and password authentication, OAuth, and JWTs.

1. Improve authentication logic in `auth.ts` using Passport.js  node middleware for standard authentication as we currently using JWTs. 
2. Integrate Passport.js middleware into API routes to handle authentication.

**Authorization:**
- Major Steps:

1. Remove scattered imperative authorization logic and replace it with calls to the centralized Casl abilities.
2. Centralize and enhance authorization logic using Casl for handling authorization logic in API side.
3. Define rules and policies using Casl to control access to resources based on the user's identity and permissions (admin, superadmin, user)
4. We can use casl in Talawa-Api and use the `packRules` and `unpackRules` function in @casl/ability/extra to transfer the necessary rules to Talawa-Admin. We can go with standard authorization logic that also supports security features and is currently the most used in the industry.
5. Document the defined authorization rules, roles, and permissions.
6. Though Oso is suitable for complex authorization logic has a longer learning curve and takes integration difficulties.

**Comment by palisadoes** [comment]
Can yoga and with be done in parallel? Is that advisable?

**Comment by varshith257** [comment]
> 3\. Can any of these steps be done in parallel?

IMO, the fastify and auth/authorization can be done parallelly.  Their logic is independent of the underlying web framework but have to ensure that the authentication and authorization mechanisms are adapted to work seamlessly with Fastify's routing system. 

For graphql-yoga there are separate mechanisms for integrating with different web frameworks. So, it could be definitely dependent issue on fastify framework. Overall, the fastify is a foundational work for these improvements of features

**Comment by palisadoes** [comment]
Can yoga and with be done in parallel? Is that advisable?

**Comment by varshith257** [comment]
> Can yoga and with be done in parallel? Is that advisable?

Definitely No, because we are trying to replace the express server. Why to go for double work them again to modify to fastify integration?

**Comment by palisadoes** [comment]
OK.

**Comment by varshith257** [comment]
I will make sure to integrate fastify framework into our project within a week

**Comment by palisadoes** [comment]
1. I've reopened the `fastify` issue.
1. @disha1202 let me know your thoughts on handling  auth

**Comment by disha1202** [comment]
Graphql-yoga is independent of express and the fastify integration will have to be updated to be made compatible with GraphQL yoga. 

Also from what I understood we will be using fastify for authentication using passport js. We'll either have to integrate express or fastify for using passport js as it is only for express-like web frameworks and not GraphQL yoga.

Auth should not be implemented in parallel to the migration. It should ideally be implemented after.

I'm still not exactly sure about this. I think we should wait for @xoldyckk inputs on this.

**Comment by varshith257** [comment]
> Graphql-yoga is independent of express and the fastify integration will have to be updated to be made compatible with GraphQL yoga. 


I think you got wrong what I Intended :
The graphql yoga has separate integrations methodology for different frameworks. More you can reference  this : https://the-guild.dev/graphql/yoga-server/docs/integrations/integration-with-fastify
https://the-guild.dev/graphql/yoga-server/docs/integrations/integration-with-express

We can start migration to graphql yoga now but we decided to replace express with fastify need to be done again work to integrate it with fastify

**Comment by varshith257** [comment]
> Also from what I understood we will be using fastify for authentication using passport js. We'll either have to integrate express or fastify for using passport js as it is only for express-like web frameworks and not GraphQL yoga.

Can work on application logic of authentication with passport.js  just the change is that routes have to adapted with compatibility of fastisfy and yoga isn't a part for this feature

**Comment by varshith257** [comment]
> Auth should not be implemented in parallel to the migration. It should ideally be implemented after.
> 

This is what I intended as of now fastify 
is foundational work for this improvement features graphql-yoga,auth and authorisation.

As @palisadoes suggested to done any work  parallely we can definitely go for authentication work as the application logic can be done within time of fastify merged  and after integration of fastify have to adapt routes according to fastify.
Though have to be waited for fastify integration

@palisadoes My suggestion is let fastisfy integration will be complete then we can work parallely with auth, authorisation and yoga. yoga can be done independently after fastify integrated 

And if auth and authorisation has to be done parallely definitely both issues has to be coordinated closely for effective auth and authorisation mechanisms

**Comment by disha1202** [comment]
> > Graphql-yoga is independent of express and the fastify integration will have to be updated to be made compatible with GraphQL yoga.
> 
> I think you got wrong what I Intended : The graphql yoga has separate integrations methodology for different frameworks. More you can reference this : https://the-guild.dev/graphql/yoga-server/docs/integrations/integration-with-fastify https://the-guild.dev/graphql/yoga-server/docs/integrations/integration-with-express
> 
> We can start migration to graphql yoga now but we decided to replace express with fastify need to be done again work to integrate it with fastify

I agree grahpql yoga has separate integrations for express and fastify. 
But rn as @xoldyckk mentioned

>> Expressjs and related packages are also mostly useless and a liability since /graphql is the only route being exposed from talawa api. So, most of the fluff related to it can also be removed.

>> If there really is a need for backend framework, some well maintained and modern framework like fastify or hono should be used. An example use case would be authentication related mechanisms like oauth openid connect which has to be handled using REST api based endpoints. 

1) Express can be completely removed and you need not to integrate fastify with apollo. 
2) We can migrate to GraphQL yoga.
3) Integrate Fastify with GraphQL yoga
4) Then start with the auth

What do you think about this?

**Comment by varshith257** [comment]
Ok, I agree with you, we can go by one by one implementing them as earlier I mentioned. 


Fastify ------> graphql-yoga -------> auth ------->authorization

**Comment by palisadoes** [comment]
Please create a GitHub discussion thread to continue the planning for this.

**Comment by pranshugupta54** [comment]
Let's continue - PalisadoesFoundation/talawa-api#1996

**Comment by varshith257** [comment]
> Let's continue - #1996

@pranshugupta54 Have created this  #1997  with current goals and plan of approach :)

**Comment by DMills27** [comment]
From my understanding, Fastify seems to have a much steeper learning curve than Express due to its plugin architecture and support for async/await. It's marketed as "faster" but in what sense? Rather than risking increasing the level of technical debt in the API, we should first verify if Fastify is indeed suitable for our use case. I suggest to code a simple mock-up of the aspects which you believe it will surpass Express in and then benchmark the stress tests for these aspects. In order words, I want to see the results of comparing Fastify vs Express. I may not be worth it in the long or short run to rewrite so much of the codebase just to accommodate a couple of milliseconds being shaved off.
